src/compiler/checker.ts

```diff
@@ -338,7 +338,13 @@ module ts {
                 // Locals of a source file are not in scope (because they get merged into the global symbol table)
                 if (location.locals && !isGlobalSourceFile(location)) {
                     if (result = getSymbol(location.locals, name, meaning)) {
-                        break loop;
+                        // Type parameters of a function are in scope in the entire function declaration, including the parameter
+                        // list and return type. However, local types are only in scope in the function body.
+                        if (!(meaning & SymbolFlags.Type) || !(result.flags & (SymbolFlags.Type & ~SymbolFlags.TypeParameter)) ||
+                            !isFunctionLike(location) || lastLocation === (<FunctionLikeDeclaration>location).body) {

```

Comment:
In general I think it is perfectly fine (and often preferable) to have multi-line boolean expressions instead of multiple `if` statements with "explaining variables". The latter just make the code even more complicated to look at. There's nothing inherently simpler about multiple `if` statements compared to multiple boolean operands separated by `&&` and `||`. Quite the opposite actually, since with boolean expressions you know there aren't any side effects or other oddities you have to worry about.

I do agree that in this case putting each condition on a separate line might make sense.